File: 0101. Symmetric Tree/solution_2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} left
 * @param {TreeNode} right
 * @return {boolean}
 */

var walkSymmetric = function(left, right)
{
    if (!left && !right)
        return true
    if (!left || !right)
        return false

    return walkSymmetric(left.left, right.right) && walkSymmetric(left.right, right.left) && left.val == right.val
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */

var isSymmetric = function(root) 
{
    if (root)
        return walkSymmetric(root.left, root.right)

    return false
};

File: 0101. Symmetric Tree/solution_2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 class Solution 
 {
     private boolean walkSymmetric(TreeNode left, TreeNode right) 
     {
         if (left == null && right == null)
             return true;
         if (left == null || right == null)
             return false;
         
         return walkSymmetric(left.left, right.right) && walkSymmetric(left.right, right.left) && left.val == right.val;
     }
     
     public boolean isSymmetric(TreeNode root) 
     {
         if (root != null)
             return walkSymmetric(root.left, root.right);
         
         return false;
     }
 }